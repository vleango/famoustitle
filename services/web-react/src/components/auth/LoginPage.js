import React, { Component, Fragment } from 'react';
import { connect } from 'react-redux';
import { Button } from 'reactstrap';
import { Form, FormGroup, Input } from 'reactstrap';
import { toastInProgress, toastSuccess, toastFail } from '../shared/Toast';

import { startLogin } from '../../actions/auth';

export class LoginPage extends Component {

    constructor(props) {
        super(props);
        this.state = {
            email: "",
            password: "",
            submitting: false,
            errorMsg: ""
        };
    }

    onInputChange = (e) => {
        const field = e.target.name;
        const value = e.target.value;
        this.setState(() => ({ [field]: value }));
    };

    onSubmitLogin = async (e) => {
        e && e.preventDefault();
        const { email, password } = this.state;
        if(email === "" || password === "") {
            this.setState({ errorMsg: "email or password is blank" });
            return;
        }

        const toastID = toastInProgress("Logging in...");

        try {
            this.setState({ submitting: true, errorMsg: "" });
            await this.props.startLogin({ email: this.state.email, password: this.state.password });
            toastSuccess("Success!", toastID);
            this.props.history.push('/');
        }
        catch (error) {
            let msg = "email and/or password was incorrect";
            if(error && error.response) {
                msg = error.response.statusText;
            }
            this.setState({ submitting: false, errorMsg: msg });
            toastFail(msg, toastID);
        }
    };

    render() {
        return (
            <Fragment>
                <div className="container">
                    <Form onSubmit={this.onSubmitLogin} autoComplete="off">
                        <FormGroup>
                            <Input type="email"
                                   name="email"
                                   value={this.state.email}
                                   placeholder="Enter your email"
                                   onChange={this.onInputChange} />
                        </FormGroup>
                        <FormGroup>
                            <Input type="password"
                                   name="password"
                                   value={this.state.password}
                                   placeholder="Enter your password"
                                   onChange={this.onInputChange} />
                        </FormGroup>

                        { this.state.errorMsg && <p>{this.state.errorMsg}</p> }

                        <div className="clearfix">
                            <Button color="primary float-right" disabled={this.state.submitting} size="lg">Login</Button>
                        </div>
                    </Form>
                </div>
            </Fragment>
        );
    }

}

const mapDispatchToProps = (dispatch) => ({
    startLogin: async (data) => await dispatch(startLogin(data))
});

export default connect(null, mapDispatchToProps)(LoginPage);
