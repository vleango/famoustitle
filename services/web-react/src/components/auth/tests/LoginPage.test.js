import React from 'react';
import { shallow } from 'enzyme';

import { LoginPage } from '../LoginPage';

describe('Components', () => {
  describe('Auth', () => {
    describe('LoginPage', () => {

      let wrapper = shallow(<LoginPage />);

      describe('Snapshot', () => {
        it('should correctly render LoginPage', () => {
          expect(wrapper).toMatchSnapshot();
        });
      });

      describe('onSubmitLogin', () => {
        describe('email is blank', () => {
          it('sets errorMsg', () => {
            wrapper.setState({ email: '', password: 'hogehoge' });
            wrapper.instance().onSubmitLogin();
            expect(wrapper.state('errorMsg')).not.toBe("");
          });
        });

        describe('password is blank', () => {
          wrapper.setState({ email: 'test@test.com', password: '' });
          wrapper.instance().onSubmitLogin();
          expect(wrapper.state('errorMsg')).not.toBe("");
        });

        describe('email and password are given', () => {
          describe('Success', () => {
            const startLogin = jest.fn(() => Promise.resolve());
            wrapper = shallow(<LoginPage startLogin={startLogin} history={[]} />);
            wrapper.setState({ email: 'test@test.com', password: 'hogehoge' });
            wrapper.instance().onSubmitLogin();
            expect(wrapper.state('submitting')).toBe(true);
          });
        });
      });

    });

  });
});
