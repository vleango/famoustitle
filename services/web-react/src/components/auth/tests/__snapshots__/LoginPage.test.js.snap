// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Components Auth LoginPage Snapshot should correctly render LoginPage 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <LoginPage />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "function",
    "props": Object {
      "children": <div
        className="container"
>
        <Form
                autoComplete="off"
                onSubmit={[Function]}
                tag="form"
        >
                <FormGroup
                        tag="div"
                >
                        <Input
                                name="email"
                                onChange={[Function]}
                                placeholder="Enter your email"
                                type="email"
                                value="test@test.com"
                        />
                </FormGroup>
                <FormGroup
                        tag="div"
                >
                        <Input
                                name="password"
                                onChange={[Function]}
                                placeholder="Enter your password"
                                type="password"
                                value=""
                        />
                </FormGroup>
                
                <div
                        className="clearfix"
                >
                        <Button
                                color="primary float-right"
                                disabled={false}
                                size="lg"
                                tag="button"
                        >
                                Login
                        </Button>
                </div>
        </Form>
</div>,
    },
    "ref": null,
    "rendered": Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": <Form
          autoComplete="off"
          onSubmit={[Function]}
          tag="form"
>
          <FormGroup
                    tag="div"
          >
                    <Input
                              name="email"
                              onChange={[Function]}
                              placeholder="Enter your email"
                              type="email"
                              value="test@test.com"
                    />
          </FormGroup>
          <FormGroup
                    tag="div"
          >
                    <Input
                              name="password"
                              onChange={[Function]}
                              placeholder="Enter your password"
                              type="password"
                              value=""
                    />
          </FormGroup>
          
          <div
                    className="clearfix"
          >
                    <Button
                              color="primary float-right"
                              disabled={false}
                              size="lg"
                              tag="button"
                    >
                              Login
                    </Button>
          </div>
</Form>,
        "className": "container",
      },
      "ref": null,
      "rendered": Object {
        "instance": null,
        "key": undefined,
        "nodeType": "function",
        "props": Object {
          "autoComplete": "off",
          "children": Array [
            <FormGroup
              tag="div"
>
              <Input
                            name="email"
                            onChange={[Function]}
                            placeholder="Enter your email"
                            type="email"
                            value="test@test.com"
              />
</FormGroup>,
            <FormGroup
              tag="div"
>
              <Input
                            name="password"
                            onChange={[Function]}
                            placeholder="Enter your password"
                            type="password"
                            value=""
              />
</FormGroup>,
            "",
            <div
              className="clearfix"
>
              <Button
                            color="primary float-right"
                            disabled={false}
                            size="lg"
                            tag="button"
              >
                            Login
              </Button>
</div>,
          ],
          "onSubmit": [Function],
          "tag": "form",
        },
        "ref": null,
        "rendered": Array [
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "function",
            "props": Object {
              "children": <Input
                name="email"
                onChange={[Function]}
                placeholder="Enter your email"
                type="email"
                value="test@test.com"
/>,
              "tag": "div",
            },
            "ref": null,
            "rendered": Object {
              "instance": null,
              "key": undefined,
              "nodeType": "class",
              "props": Object {
                "name": "email",
                "onChange": [Function],
                "placeholder": "Enter your email",
                "type": "email",
                "value": "test@test.com",
              },
              "ref": null,
              "rendered": null,
              "type": [Function],
            },
            "type": [Function],
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "function",
            "props": Object {
              "children": <Input
                name="password"
                onChange={[Function]}
                placeholder="Enter your password"
                type="password"
                value=""
/>,
              "tag": "div",
            },
            "ref": null,
            "rendered": Object {
              "instance": null,
              "key": undefined,
              "nodeType": "class",
              "props": Object {
                "name": "password",
                "onChange": [Function],
                "placeholder": "Enter your password",
                "type": "password",
                "value": "",
              },
              "ref": null,
              "rendered": null,
              "type": [Function],
            },
            "type": [Function],
          },
          "",
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": <Button
                color="primary float-right"
                disabled={false}
                size="lg"
                tag="button"
>
                Login
</Button>,
              "className": "clearfix",
            },
            "ref": null,
            "rendered": Object {
              "instance": null,
              "key": undefined,
              "nodeType": "class",
              "props": Object {
                "children": "Login",
                "color": "primary float-right",
                "disabled": false,
                "size": "lg",
                "tag": "button",
              },
              "ref": null,
              "rendered": "Login",
              "type": [Function],
            },
            "type": "div",
          },
        ],
        "type": [Function],
      },
      "type": "div",
    },
    "type": Symbol(react.fragment),
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "function",
      "props": Object {
        "children": <div
          className="container"
>
          <Form
                    autoComplete="off"
                    onSubmit={[Function]}
                    tag="form"
          >
                    <FormGroup
                              tag="div"
                    >
                              <Input
                                        name="email"
                                        onChange={[Function]}
                                        placeholder="Enter your email"
                                        type="email"
                                        value="test@test.com"
                              />
                    </FormGroup>
                    <FormGroup
                              tag="div"
                    >
                              <Input
                                        name="password"
                                        onChange={[Function]}
                                        placeholder="Enter your password"
                                        type="password"
                                        value=""
                              />
                    </FormGroup>
                    
                    <div
                              className="clearfix"
                    >
                              <Button
                                        color="primary float-right"
                                        disabled={false}
                                        size="lg"
                                        tag="button"
                              >
                                        Login
                              </Button>
                    </div>
          </Form>
</div>,
      },
      "ref": null,
      "rendered": Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": <Form
            autoComplete="off"
            onSubmit={[Function]}
            tag="form"
>
            <FormGroup
                        tag="div"
            >
                        <Input
                                    name="email"
                                    onChange={[Function]}
                                    placeholder="Enter your email"
                                    type="email"
                                    value="test@test.com"
                        />
            </FormGroup>
            <FormGroup
                        tag="div"
            >
                        <Input
                                    name="password"
                                    onChange={[Function]}
                                    placeholder="Enter your password"
                                    type="password"
                                    value=""
                        />
            </FormGroup>
            
            <div
                        className="clearfix"
            >
                        <Button
                                    color="primary float-right"
                                    disabled={false}
                                    size="lg"
                                    tag="button"
                        >
                                    Login
                        </Button>
            </div>
</Form>,
          "className": "container",
        },
        "ref": null,
        "rendered": Object {
          "instance": null,
          "key": undefined,
          "nodeType": "function",
          "props": Object {
            "autoComplete": "off",
            "children": Array [
              <FormGroup
                tag="div"
>
                <Input
                                name="email"
                                onChange={[Function]}
                                placeholder="Enter your email"
                                type="email"
                                value="test@test.com"
                />
</FormGroup>,
              <FormGroup
                tag="div"
>
                <Input
                                name="password"
                                onChange={[Function]}
                                placeholder="Enter your password"
                                type="password"
                                value=""
                />
</FormGroup>,
              "",
              <div
                className="clearfix"
>
                <Button
                                color="primary float-right"
                                disabled={false}
                                size="lg"
                                tag="button"
                >
                                Login
                </Button>
</div>,
            ],
            "onSubmit": [Function],
            "tag": "form",
          },
          "ref": null,
          "rendered": Array [
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "function",
              "props": Object {
                "children": <Input
                  name="email"
                  onChange={[Function]}
                  placeholder="Enter your email"
                  type="email"
                  value="test@test.com"
/>,
                "tag": "div",
              },
              "ref": null,
              "rendered": Object {
                "instance": null,
                "key": undefined,
                "nodeType": "class",
                "props": Object {
                  "name": "email",
                  "onChange": [Function],
                  "placeholder": "Enter your email",
                  "type": "email",
                  "value": "test@test.com",
                },
                "ref": null,
                "rendered": null,
                "type": [Function],
              },
              "type": [Function],
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "function",
              "props": Object {
                "children": <Input
                  name="password"
                  onChange={[Function]}
                  placeholder="Enter your password"
                  type="password"
                  value=""
/>,
                "tag": "div",
              },
              "ref": null,
              "rendered": Object {
                "instance": null,
                "key": undefined,
                "nodeType": "class",
                "props": Object {
                  "name": "password",
                  "onChange": [Function],
                  "placeholder": "Enter your password",
                  "type": "password",
                  "value": "",
                },
                "ref": null,
                "rendered": null,
                "type": [Function],
              },
              "type": [Function],
            },
            "",
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": <Button
                  color="primary float-right"
                  disabled={false}
                  size="lg"
                  tag="button"
>
                  Login
</Button>,
                "className": "clearfix",
              },
              "ref": null,
              "rendered": Object {
                "instance": null,
                "key": undefined,
                "nodeType": "class",
                "props": Object {
                  "children": "Login",
                  "color": "primary float-right",
                  "disabled": false,
                  "size": "lg",
                  "tag": "button",
                },
                "ref": null,
                "rendered": "Login",
                "type": [Function],
              },
              "type": "div",
            },
          ],
          "type": [Function],
        },
        "type": "div",
      },
      "type": Symbol(react.fragment),
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
  },
}
`;
